changelogs/unreleased/gh-4693-formats-for-spaceless-tuples.md

```diff
@@ -0,0 +1,4 @@
+## feature/box
```

Comment:
This commit has a lot of changes in different parts of box. It looks far away from being atomic. I suggest to split it to meaningful parts: add necessary functionality on a lower level, then add it to the upper level, etc.

Reply:
Could you please elaborate which parts you don't consider atomic?

Reply:
* Add `is_external` and related API change in `box/format.c` (and needed changes at the caller side).
  - Hmm. Just found that I don't know a reason why the new flag is needed. We don't change any logic here, right? And correct runtime formats were created before.
* Prepare for possible lack of` fkey_same_space` (BTW, where it was introduced?).
* Tuple constrains encoding.
* Tuple format encoding.
* Some format comparison function change (?).
* The default value and expressions changes.
  - BTW, are those default value now working for all the formats? Can I set a default value for a field that will be autoadded to a space tuple at insert/replace?
  - And why it is required here? An extra field 'is needed for tuple format
serialization' looks as a strange motivation. Why not two more fields?
* lua/format: rewrite existing tuple format functions from cdata to userdata.
* lua/format: add public tuple format Lua constructor (plus __gc hook).
* lua/format: add __serialize for tuple format objects.
* lua/format: add pairs/ipairs for tuple format objects.
* lua/tuple: add options and the format option into `box.tuple.new()`.

As you see, I'm not very familiar with tuple format internals :)

As you see, I don't understand the 3rd patch fully, so the splitting is quite arbitrary. OTOH, more granular changes would make it more clear for me and easier to review/read in a future by others.